### PR TITLE
Implement gulp build system

### PR DESCRIPTION
**Pull request by sweepyoface**
**This build process:**
- Minifies and compiles SASS to CSS
- Minifies JS
- Minifies HTML
- Automatically adds a query string of the file's hash to assets to invalidate browser caches

**To build:**
`yarn && yarn run build`

Thanks to @kashike for some gulpfile code and layout template.

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+const gulp = require("gulp");
+const gulp_htmlmin = require("gulp-htmlmin");
+const gulp_sass = require("gulp-sass");
+const gulp_terser = require("gulp-terser");
+const gulp_twig = require("gulp-twig");
+const gulp_buster = require("gulp-buster");
+const gulp_rename = require("gulp-rename");
+
+gulp.task("sass", function() {
+	return gulp
```
Comment (kashike) [inline gulpfile.js]: no tabs.

File: gulpfile.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	return new Promise((resolve, reject) => {
+		let cssBusters = JSON.parse(fs.readFileSync("dist/css/busters.json"));
+		let jsBusters = JSON.parse(fs.readFileSync("dist/js/busters.json"));
+		let imageBusters = JSON.parse(fs.readFileSync("dist/images/busters.json"));
+
+		gulp.src("src/*.twig")
+			.pipe(
+				gulp_twig({
+					data: {
+						community_css_version: cssBusters["dist/css/community.css"],
```
Comment (aikar) [inline gulpfile.js]: Can we automate this in some way to precompute these.

Comment (sweepyoface) [reply gulpfile.js]: Couldn't think of a good way to do it. Any ideas?

**Comment by aikar** [comment]
Should be able to use a glob library and find all static files. Why does it even need a local file. It should calculate a hash based on the content

**Comment by sweepyoface** [comment]
> That file has no Twig logic. I'm actually going to use a single layout file instead with blocks for page-specific content (the pages will extend it), so it'll be the only include.

Scratch that, I'm getting rid of all includes.

**Comment by sweepyoface** [comment]
> Should be able to use a glob library and find all static files. Why does it even need a local file. It should calculate a hash based on the content

Not sure what you mean here. `gulp-buster` *does* calculate hashes when given a glob, the issue is applying them to HTML references with Twig. Since it's in `{"path":"hash"}` format I suppose we could match the path (after `dist/`) but that would require some regex.

File: src/partial/layout.twig [hunk 3]
```diff
@@ -39,60 +40,83 @@
...
+          <a data-target="sidenav" class="hide-on-large-only white-text button-collapse sidenav-button sidenav-trigger clickable">
+            <i class="material-icons">menu</i>
+          </a>
           <ul class="right hide-on-med-and-down">
             <!--li><a class="grey-text text-darken-4 btn btn-flat waves-effect" href="https://papermc.io/javadocs/"><i class="material-icons left">info</i>JAVADOCS</a></li-->
-            <li><a class="white-text btn btn-flat waves-effect" href="community"><i class="material-icons left">people</i>COMMUNITY</a></li>
-            <li><a class="white-text btn btn-flat waves-effect" href="https://github.com/PaperMC/Paper"><i class="material-icons left">code</i>GITHUB</a></li>
-            <li><a class="white-text btn btn-flat waves-effect" href="https://paper.readthedocs.io/"><i class="material-icons left">help_outline</i>DOCUMENTATION</a></li>
-            <li><a class="downloads-button white grey-text text-darken-4 btn nav-btn waves-effect" href="downloads"><i class="material-icons left">cloud_download</i>Downloads</a></li>
+            <li>
```
Comment (aikar) [inline src/partial/layout.twig]: Formatting got weird here. if we're gonna move <a> to its own line, </a> needs to be on its own line.

File: gulpfile.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  return gulp
+    .src("src/favicons/*")
+    .pipe(gulp.dest("dist/favicons"))
+})
+
+gulp.task("bust", function() {
+  return Promise.all([
+    // CSS busters
+    new Promise((resolve, reject) => {
+      gulp.src("dist/css/*.css")
```
Comment (aikar) [inline gulpfile.js]: This could just be a single bust as `"dist/**/*"`
anything in dist should have a bust entry

File: gulpfile.js [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+gulp.task("content", function() {
+  return new Promise((resolve, reject) => {
+    let cssBusters = JSON.parse(fs.readFileSync("dist/css/busters.json"));
+    let jsBusters = JSON.parse(fs.readFileSync("dist/js/busters.json"));
+    let imageBusters = JSON.parse(fs.readFileSync("dist/images/busters.json"));
+
+    gulp.src("src/*.twig")
+      .pipe(
+        gulp_twig({
+          data: {
```
Comment (aikar) [inline gulpfile.js]: just pass this as a single object as I mentioned above

data: {
  cache: busters
}

then templates can access v={{ cache['css/blah.css'] }}

File: src/partial/layout.twig (hunk 3 above)
Comment (sweepyoface) [reply src/partial/layout.twig]: Done, not sure why my editor does that.

**Comment by sweepyoface** [comment]
Should be all automatic now, with the variable being the file basename with `.` replaced by `_` and `-` replaced by `__`.

**Comment by aikar** [comment]
It works as is, but curious why the rewrite? polluting the global variable map feels weird for the cache ID's, as it could unintentionally overwrite something

I do like the building of the url, but can't we do 

`<script src="{{ urls['js/script.js' }}"></script>`

Feels more natural to me that you can have 1 root data object that has all the urls under it. no risk of collusions, clean and intuitive on what the syntax should be

as it is with the rewriting, to add a new js/css you gotta know the rewrite rules too (- -> __ etc), which is unnecessary.

If you can update that, I think I'm good

**Comment by sweepyoface** [comment]
I don't think Twig supports objects like that? They only support dot notation.

**Comment by aikar** [comment]
my suggestion for @zachbr as said on discord, suggested release policy would be something like

```bash
#!/usr/bin/env bash
set -e
build=$(date -Is)
reldir=`pwd`/../releases/
echo "Building site $build"
npm run build
echo "Build success"
mv dist $reldir/$build;
(
    cd $reldir
    echo "Current release is"
    ls -la | grep current
    test -d $build && ln -sfn $build current 
    echo "Updated current symlink to $build. Site is live"
)
```
then point webserver at current symlink

**Pull request merged**